Add ShouldContinue for removing folder

quoctruong opened the pull request:
Also, use path separator based on OS.

![image](https://cloud.githubusercontent.com/assets/4371645/23186981/ec104004-f83d-11e6-80d7-80166c596493.png)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/googlecloudplatform/google-cloud-powershell/467)
<!-- Reviewable:end -->

ILMTitan commented on Google.PowerShell/Provider/GoogleCloudStorageProvider.cs:
```diff
@@ -859,13 +861,32 @@ protected override object RemoveItemDynamicParameters(string path, bool recurse)
+            // Ask the user for permission to delete children of the folder if -Recurse is not supplied.
+            if (children.Any() && !recurse)
+            {
+                if (ShouldContinue(query, caption))
```
If we want to force this, we should delete the children first in a loop checked by the four parameter ShouldContinue. This will let the user examine the names of each GcsObject they are deleting, or just say "Yes to All".

ILMTitan commented on Google.PowerShell/Provider/GoogleCloudStorageProvider.cs:
```diff
@@ -835,6 +836,7 @@ protected override void RemoveItem(string path, bool recurse)
                 case GcsPath.GcsPathType.Object:
                     if (IsItemContainer(path))
                     {
+                        path = path.TrimEnd("/\\".ToCharArray());
```
String.TrimEnd can take a params array. This could be `path.TrimEnd('/', '\\');`

ILMTitan commented on Google.PowerShell/Provider/GoogleCloudStorageProvider.cs:
```diff
@@ -859,13 +861,32 @@ protected override object RemoveItemDynamicParameters(string path, bool recurse)
+                }
+                else
+                {
+                    return;
```
Even if we don't delete all of the children, we should try to delete the folder object if it is real. It could contain undesired data and there is already a ShouldProcess gate on it.

quoctruong commented:
@ILMTitan PTAL. Turns out there is a bug in BucketModel that makes the provider fails to detect whether a folder has children or not. When I fixed the bug, we don't need the extra logic in RemoveFolder so I reverted that.

ILMTitan commented on Google.PowerShell/Provider/BucketModel.cs:
```diff
@@ -77,7 +77,9 @@ public void PopulateModel()
                 string prefix = gcsObject.Name.TrimEnd(Separator);
                 int lastSeparator = gcsObject.Name.LastIndexOf(Separator);
-                bool children = false;
+                // If the object name is testing/blah.txt, then it has children since / is not at the end.
```
Consider changing "it" to "the first prefix". "It" sounds like you are talking about "testing/blah.txt", which probably does not have children.

quoctruong commented:
Thanks!
